bd-session-replay/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+#[cfg(test)]
+#[path = "./recorder_test.rs"]
+mod recorder_test;
+
+use bd_runtime::runtime::{session_replay, BoolWatch, ConfigLoader, DurationWatch};
+use bd_shutdown::{ComponentShutdown, ComponentShutdownTrigger};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::{Instant, Interval};
+
+#[cfg(test)]
+#[ctor::ctor]
+fn test_global_init() {
+  bd_test_helpers::test_global_init();
+}
+
+//
+// Target
+//
+
+// An interface implementing the act of capturing user screens.
+pub trait Target {
+  // Instrument the target to capture a privacy-preserving and bandwidth-efficient representation
+  // of the user's screen. The target should capture the wireframe and send it using the
+  // `logger::log_session_replay_wireframe` method. The target is expected to operate
+  // asynchronously, as accessing the application view hierarchy requires executing on the main
+  // thread, and the `capture_wireframe` method is always called from a non-main thread.
+  fn capture_wireframe(&self);
+
+  // Instrument the target to capture a pixel-perfect representation of the user's screen.
+  // The target should capture the wireframe and send it using the
+  // `logger::log_session_screenshot` method. The target is expected to operate asynchronously,
+  // as accessing the application view hierarchy requires executing on the main thread, and the
+  // `take_screenshot` method is always called from a non-main thread.
+  fn take_screenshot(&self);
+}
+
+//
+// Recorder
+//
+
+pub struct Recorder {
+  target: Box<dyn Target + Send + Sync>,
+
+  is_periodic_reporting_enabled: bool,
+  reporting_interval_rate: Duration,
+  reporting_interval: Option<Interval>,
+
+  is_screenshotting_enabled: bool,
+  take_screenshot_rx: Receiver<()>,
+
+  is_periodic_reporting_enabled_flag: BoolWatch<session_replay::PeriodicWireframesEnabledFlag>,
+  reporting_interval_flag: DurationWatch<session_replay::ReportingIntervalFlag>,
+  is_screenshotting_enabled_flag: BoolWatch<session_replay::ScreenshotsEnabledFlag>,
+}
+
+impl Recorder {
+  pub fn new(
+    target: Box<dyn Target + Send + Sync>,
+    runtime_loader: &Arc<ConfigLoader>,
+  ) -> (Self, Sender<()>) {
+    // Limit the buffer size to 1 to reduce the risk of putting too much pressure on the
+    // application's main thread when dequeuing the screenshot actions from the channel and
+    // passing them to the platform layer, which performs the actual screenshotting on the main
+    // thread.
```

Comment:
Sorry for possibly asking the obvious, but will the platform layer block until the screen shot is taken? If it spawns some other async task this would still be able to loop unbounded.